bin/GoogleAdsCleanupServices.php

```diff
@@ -0,0 +1,529 @@
+		$this->remove_pattern( $file, $pattern );
+	}
+
+	/**
+	 * Remove a function from a file.
+	 *
+	 * @param string $file
+	 * @param string $name Function name.
+	 */
+	protected function remove_function( string $file, string $name ) {
+		$file = $this->file_path( $file );
+		if ( ! file_exists( $file ) ) {
+			$this->warning( sprintf( 'File does not exist: %s', $file ) );
+			return;
+		}
+
+		$contents = file_get_contents( $file ); // phpcs:ignore WordPress.WP.AlternativeFunctions
+
+		$pattern = '/function ' . preg_quote( $name, '/' ) . '\(/';
+		if ( ! preg_match( $pattern, $contents, $matches, PREG_OFFSET_CAPTURE ) ) {
+			$this->warning( sprintf( 'Function %s not found in %s', $name, $file ) );
+			return;
+		}
+
+		$offset  = $matches[0][1];
+		$length  = strlen( $contents );
+		$bracket = 0;
+
+		// Parse until we find beginning of doc block.
+		$start = strrpos( $contents, '/**', ( $length - $offset ) * -1 );
+
+		// Parse until we encounter closing bracket.
+		for ( $end = $offset; $end < $length; $end++ ) {
+			if ( '{' === $contents[ $end ] ) {
+				$bracket++;
+			}
+			if ( '}' === $contents[ $end ] ) {
+				$bracket--;
+				if ( 1 > $bracket ) {
+					break;
+				}
+			}
+		}
+
+		if ( false === $start || 0 > $start || $end >= $length ) {
+			$this->warning( sprintf( 'Function %s not found in %s', $name, $file ) );
+			return;
+		}
+
+		// Include whitespaces before start.
+		while ( 0 < $start && ctype_space( $contents[ $start - 1 ] ) ) {
+			$start--;
+		}
+
+		$new  = substr( $contents, 0, $start );
+		$new .= substr( $contents, $end + 1 );
+
+		if ( empty( $new ) ) {
+			$this->warning( sprintf( 'Replace failed for function %s in %s', $name, $file ) );
+			return;
+		}
+
+		file_put_contents( $file, $new ); // phpcs:ignore WordPress.WP.AlternativeFunctions
+	}
+
+	/**
+	 * Remove regex pattern from a file.
+	 *
+	 * @param string $file
+	 * @param string $pattern Regex pattern to match.
+	 */
+	protected function remove_pattern( string $file, string $pattern ) {
+		$file = $this->file_path( $file );
+		if ( ! file_exists( $file ) ) {
+			$this->warning( sprintf( 'File does not exist: %s', $file ) );
+			return;
+		}
+
+		$contents = file_get_contents( $file ); // phpcs:ignore WordPress.WP.AlternativeFunctions
+		if ( ! preg_match( $pattern, $contents, $matches ) ) {
+			$this->warning( sprintf( 'Pattern %s not found in %s', $pattern, $file ) );
+			return;
+		}
+
+		$new = preg_replace( $pattern, '', $contents );
+		if ( empty( $new ) ) {
+			$this->warning( sprintf( 'Replace failed for pattern %s in %s', $pattern, $file ) );
+			return;
+		}
+
+		file_put_contents( $file, $new ); // phpcs:ignore WordPress.WP.AlternativeFunctions
+	}
+
+	/**
+	 * Remove a file.
+	 *
+	 * @param string $file
+	 */
+	protected function remove_file( string $file ) {
+		$file = $this->file_path( $file );
+		if ( ! file_exists( $file ) ) {
+			$this->warning( sprintf( 'File does not exist: %s', $file ) );
+			return;
+		}
```

Comment:
💅 

Duplicated code in 445

Reply:
I personally think since this is a helper script and not included in the final code. It's an ok balance to have this duplicated code instead of abstracting everything into separate functions. Since we aren't using exception handling even if a helper function would be used it would still need to check a condition to return early, so there isn't a lot of duplication you can prevent here.